cmd/ipi-deprovision/ipi-deprovision.sh

```diff
@@ -4,64 +4,60 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-### cluster_age_cutoff is going to be used as argument in the aws command below
-### to filter out the expired clusters with the tags of the key expirationDate
-### which is always with timezone +00:00, eg, "2019-07-23T22:35+0000".
-### CLUSTER_TTL controls the relative time to the running time of this script.
-### eg, export CLUSTER_TTL="22 hours ago"
+# CLUSTER_TTL controls the relative time to the running time of this script.
+# eg, export CLUSTER_TTL="22 hours ago"
 echo "Searching for clusters with a TTL of ${CLUSTER_TTL}"
 cluster_age_cutoff="$(TZ=":Africa/Abidjan" date --date="${CLUSTER_TTL}" '+%Y-%m-%dT%H:%M+0000')"
 
 echo "cluster_age_cutoff: ${cluster_age_cutoff}"
 
-### eg, export AWS_REGIONS="us-east-1;us-west-1"
-echo "take regions from env. var.: AWS_REGIONS: ${AWS_REGIONS}"
-IFS=';' read -r -a regions <<< "$AWS_REGIONS"
-
-handle_cluster () {
-  local cluster_name
-  cluster_name=$1
-  local expirationDateValue
-  expirationDateValue=$2
-  local region
-  region=$3
-  echo "handling cluster: ${cluster_name} in region ${region} with expirationDate: ${expirationDateValue}"
-  timeout 10m ./bin/hiveutil aws-tag-deprovision --loglevel debug "${cluster_name}=owned" --region "${region}"
+echo "deprovisioning in AWS ..."
+for region in $( aws ec2 describe-regions --query "Regions[].{Name:RegionName}" --output text ); do
+  echo "deprovisioning in AWS region ${region} ..."
+  for cluster in $( aws ec2 describe-vpcs --output json --region "${region}" | jq --arg date "${cluster_age_cutoff}" -r -S '.Vpcs[] | select (.Tags[]? | (.Key == "expirationDate" and .Value < $date)) | .Tags[] | select (.Value == "owned") | .Key' | shuf ); do
+    workdir="/tmp/deprivison/aws/${infraID}"
+    mkdir -p "${workdir}"
+    cat <<EOF >"${workdir}/metadata.json"
+{
+  "aws":{
+    "region":"${region}",
+    "identifier":[{
+      "${cluster}": "owned"
+    }]
+  }
 }
+EOF
+    echo "will deprivison AWS cluster ${cluster} in region ${region}:"
+    cat "${workdir}/metadata.json"
+  done
+done
 
-collect_metadata () {
-  local cluster_name
-  cluster_name=$1
-  ###eg cluster_name=kubernetes.io/cluster/ci-op-724qy8fn-55c01-s5c8w
-  local ns
-  ns=${cluster_name:22:14}
-  echo "cluster is used by namespace ${ns}"
-  if ! oc get ns "${ns}" >/dev/null 2>&1; then
-    echo "namespace ${ns} does not exist any more"
-    return
+echo "deprovisioning in GCE ..."
```

Comment:
Running GCE after AWS will expose us to GCE leaks if AWS gets stuck.  Compare openshift/ci-tools#231 where we worked around AWS-side teardown errors blocking us from removing other leaked AWS clusters.  I'd rather run the AWS and GCP portions in parallel (in separate scripts?  Based on a argument passed to the script?  In parallel in the one script via backgrounded functions?).

But as long as we don't get stuck on anything, serial is still better than "no GCP reaping at all".

Reply:
Fixed via #387.